src/NLog/Targets/DatabaseTarget.cs

```diff
@@ -790,40 +833,137 @@ private void RunInstallCommands(InstallationContext installationContext, IEnumer
         {
-            for(int i = 0; i < databaseParameterInfos.Count; ++i)
+            for (int i = 0; i < databaseParameterInfos.Count; ++i)
             {
                 DatabaseParameterInfo par = databaseParameterInfos[i];
-                IDbDataParameter p = command.CreateParameter();
-                p.Direction = ParameterDirection.Input;
+                IDbDataParameter dbParameter = command.CreateParameter();
+                EnsureResolveParameterInfo(command, dbParameter);
+                dbParameter.Direction = ParameterDirection.Input;
                 if (par.Name != null)
                 {
-                    p.ParameterName = par.Name;
+                    dbParameter.ParameterName = par.Name;
                 }
 
                 if (par.Size != 0)
                 {
-                    p.Size = par.Size;
+                    dbParameter.Size = par.Size;
                 }
 
                 if (par.Precision != 0)
                 {
-                    p.Precision = par.Precision;
+                    dbParameter.Precision = par.Precision;
                 }
 
                 if (par.Scale != 0)
                 {
-                    p.Scale = par.Scale;
+                    dbParameter.Scale = par.Scale;
                 }
 
-                string stringValue = RenderLogEvent(par.Layout, logEvent);
+                _parameterTypeSetter.SetParameterDbType(dbParameter, par);
+                object value;
+
+                try
+                {
+                    value = GetParameterValue(logEvent, par, dbParameter.ParameterName, dbParameter.DbType);
+                }
+                catch (Exception exception)
+                {
+                    // enhance context. Already checked for must-rethrow 
+                    exception.Data["parameter"] = dbParameter;
+                    throw;
+                }
+
+
+                dbParameter.Value = value;
+                InternalLogger.Trace("  DatabaseTarget: Parameter: '{0}' = '{1}' ({2})", dbParameter.ParameterName, value, dbParameter.DbType);
+                command.Parameters.Add(dbParameter);
 
-                p.Value = stringValue;
-                command.Parameters.Add(p);
 
-                InternalLogger.Trace("  DatabaseTarget: Parameter: '{0}' = '{1}' ({2})", p.ParameterName, p.Value, p.DbType);
             }
         }
 
+        /// <summary>
+        /// Get (converted) parameter value, including fallbacks
+        /// </summary>
+        /// <param name="logEvent">Current logevent.</param>
+        /// <param name="parameterInfo">Parameter config</param>
+        /// <param name="parameterName"></param>
+        /// <param name="dbType"></param>
+        /// <returns></returns>
+        private object GetParameterValue(LogEventInfo logEvent, DatabaseParameterInfo parameterInfo, string parameterName, DbType dbType)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)